cmd/promoted-image-governor/main.go

```diff
@@ -262,6 +272,47 @@ func isMirroredFromOCP(tag api.ImageStreamTagReference, refs []ImageStreamRef) b
 	return false
 }
 
+type githubClient interface {
+	GetRef(org, repo, ref string) (string, error)
+}
+
+// checkImageStreamTags checks if promotedTags are built with the current head of org/repo/branch
+func checkImageStreamTags(ctx context.Context, client ctrlruntimeclient.Client, promotedTags map[api.ImageStreamTagReference]api.Metadata, gitHubClient githubClient) []error {
+	var errs []error
+	for promotedTag, metadata := range promotedTags {
+		ist := &imagev1.ImageStreamTag{}
+		istName := fmt.Sprintf("%s:%s", promotedTag.Name, promotedTag.Tag)
+		if err := client.Get(ctx, ctrlruntimeclient.ObjectKey{Namespace: promotedTag.Namespace, Name: istName}, ist); err != nil {
+			if kerrors.IsNotFound(err) {
+				// The tag has not been promoted yet
+				logrus.WithError(err).Warnf("failed to get promoted isTag %s in namespace %s", istName, promotedTag.Namespace)
+				continue
+			}
+			errs = append(errs, fmt.Errorf("failed to get promoted isTag %s in namespace %s: %w", istName, promotedTag.Namespace, err))
+			continue
+		}
+		istCommit, err := promotionreconciler.CommitForIST(ist)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("failed to get commit for isTag %s in namespace %s: %w", istName, promotedTag.Namespace, err))
+			continue
+		}
+		log := logrus.WithField("org", metadata.Org).WithField("repo", metadata.Repo).WithField("branch", metadata.Branch)
+		currentHEAD, found, err := promotionreconciler.CurrentHEADForBranch(gitHubClient, metadata, log)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("failed to get current git head for isTag %s in namespace %s: %w", istName, promotedTag.Namespace, err))
+			continue
+		}
+		if !found {
+			errs = append(errs, fmt.Errorf("got 404 for %s/%s/%s from github for isTag %s in namespace %s, this likely means the repo or branch got deleted or we are not allowed to access it", metadata.Org, metadata.Repo, metadata.Branch, istName, promotedTag.Namespace))
+			continue
+		}
+		if currentHEAD != istCommit {
+			errs = append(errs, fmt.Errorf("the isTag %s in namespace %s is not built from the head %s of %s/%s/%s", istName, promotedTag.Namespace, currentHEAD, metadata.Org, metadata.Repo, metadata.Branch))
```

Comment:
We should just trigger the job instead of logging something

Reply:
I'm also pretty sure I wrote a postsubmit for `openshift/release` that did this in the past, but it gor reverted due to some issues - can you look that up?

Reply:
Probably the postsubmit has been replaced by https://github.com/openshift/ci-tools/tree/master/pkg/controller/promotionreconciler

Back in time, promotionreconciler stopped working for some reason.
The current PR for DPTP-1899 will detect it and fire alerts (since it is an infra job).